### PR TITLE
sql: fix some issues around the portal state management

### PR DESCRIPTION
**Pull request by yuzefovich**
This commit fixes some issues around the handling of portals in the
prepared statements namespace. Namely:
- it fixes the reference counting that was broken in cf0d5e4 (where we
switched to storing the portal by value instead of by pointer, but the
refcounts need to be shared between all references which wasn't true
in that commit) by entirely removing the ref counting
- it fixes the missing memory accounting for a copy of the portal
- it deletes all of the portals from `prepStmtsNamespaceAtTxnRewindPos`
in `resetExtraTxnState` whenever a txn commits or rolls back
- it adjusts a comment on the method deleting a prepared statement to
match the implementation.

Release note: None.

Release justification: fix to a long-standing bug.

### CODE REVIEW

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

Consider embedding `NoCopy` in `PreparedStatement` and `PreparedPortal` (you can mention in a comment that we do refcounting so we don't want to allow copy-by-value).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-:-Mi82gkvD8vfvutAHITz:b-rlqzzm)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei and @rafiss)

**Review by andreimatei — COMMENTED** [review]
It's been too long for me, so I'm not gonna look at this. Sorry for my past sins, if any.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-:-Mi91Wk7DfKs5Cq5BdJh:bj60tuw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

File: pkg/sql/conn_executor.go
Comment (rafiss) [inline pkg/sql/conn_executor.go:1469]: i noticed this comment on `prepStmtsNamespaceAtTxnRewindPos`:
> Prepared statements are shared between the two collections, but these collections are periodically reconciled.

i wonder if we lose track of a reference when that happens?

**Review by rafiss — APPROVED** [review]
lgtm! my comment is just a vague idea about another place to look into

**Review by yuzefovich — COMMENTED** [review]
The initial version of this PR was broken, so I adjusted the first commit to truly fix the reference counting issue without introducing regressions. I'm currently running the `schemachange` roachtest with the build statement cache re-enabled, and it looks like it is passing (the RSS stays very reasonable), so the incorrect reference counting was the root cause of those OOMs.

Here is a bit of background on why the first commit is as it is: we have somewhat conflicting goals when handling `PreparedPortal` struct - namely, the `refCount` object needs to be shared about all references to the prepared portals that were created from the same prepared statement while the `exhausted` field needs to be separate for each instance of the prepared portal. The code on master breaks the former requirement (we're performing an incorrect reference counting), the original version of this PR broke the latter requirement. I don't see another way from storing `PreparedPortal` as a combination of "by pointer" (for the former) and "by value" (for the latter), so this commit keeps handling the whole struct by value yet it extracts the ref counter to be by pointer. Please let me know what you think.

I also adjusted the deletion of portals from `prepStmtsNamespaceAtTxnRewindPos` to be done in `resetExtraTxnState` only for commits and rollbacks (and not for restarts). However, this doesn't seem to have influence on passing of the `schemachange` roachtest, so I'm not certain that we should keep it or not. I'm leaning towards keeping it in since it seems like a beneficial change overall.

@raduberinde @rafiss please take another look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-:-MiDJbqU4s4NXSYayyQj:b-ta1xvb)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss)

___
*[pkg/sql/conn_executor.go, line 1469 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-Mi7eU0HsQLYTpdMNDNP-r1-1469:-MiDLHWi4zGLbNBZXgTl:b-7pc4e5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6117699bba790bed5b5fed068859b3324b3df341/pkg/sql/conn_executor.go#L1469)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i noticed this comment on `prepStmtsNamespaceAtTxnRewindPos`:
> Prepared statements are shared between the two collections, but these collections are periodically reconciled.

i wonder if we lose track of a reference when that happens?
</blockquote></details>

I don't think so, see my explanation in the main thread.

File: pkg/workload/pgx_helpers.go
Comment (rafiss) [inline pkg/workload/pgx_helpers.go:66]: i vote that we should still keep this cache disabled, but curious to hear any thoughts on this. i don't feel too strongly, but my reasoning is just that it's not that obvious to a writer of a workload that this statement cache will cause all statements to be prepared. but perhaps that doesn't matter so much.

File: pkg/sql/prepared_stmt.go
Comment (rafiss) [inline pkg/sql/prepared_stmt.go:136]: thanks for explaining in the PR! that makes me wonder... maybe we could just keep using the `PreparedPortal` by value (and somehow require that it's only used by value). it seems like we only need to care about counting references to `Stmt  *PreparedStatement`. so would things work if we deleted `mutablePreparedPortalState` entirely and made this update instead?

```
-// decRef decrements the number of references to this portal. If the refCount
-// reaches 0, then the memory account is shrunk accordingly.
+// decRef decrements the number of references to the prepared statement used
+// by this portal. The memory account is shrunk accordingly.
 func (p *PreparedPortal) decRef(
     ctx context.Context, prepStmtsNamespaceMemAcc *mon.BoundAccount, portalName string,
 ) {
-    if p.refCount <= 0 {
-        log.Fatal(ctx, "corrupt PreparedPortal refcount")
-    }
-    p.refCount--
-
-    if p.refCount == 0 {
-        prepStmtsNamespaceMemAcc.Shrink(ctx, p.size(portalName))
-        p.Stmt.decRef(ctx)
-    }
+    prepStmtsNamespaceMemAcc.Shrink(ctx, p.size(portalName))
+    p.Stmt.decRef(ctx)
 }
```

**Review by RaduBerinde — COMMENTED** [review]
Can you explain how the incorrect reference counting led to OOMs? The reference counting only seems to determine whether we close the `memAcc`, and not whether we still keep a `*PreparedStatement` somewhere. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-:-MiNTrWF8z2gXorwAuI9:bvfvyek)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss and @yuzefovich)

___
*[pkg/sql/conn_executor.go, line 1476 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-MiNUTu_-5V0z3rNX6qG:-MiNUTu_-5V0z3rNX6qH:b9qn7aq) ([raw file](https://github.com/cockroachdb/cockroach/blob/92142a550d4f97509aaad1a6ee0c3f8edef831fe/pkg/sql/conn_executor.go#L1476)):*
> ```Go
> 		for name, p := range ex.extraTxnState.prepStmtsNamespaceAtTxnRewindPos.portals {
> 			p.decRef(ctx, &ex.extraTxnState.prepStmtsNamespaceMemAcc, name)
> 			delete(ex.extraTxnState.prepStmtsNamespaceAtTxnRewindPos.portals, name)
> ```

Is this part what fixes the OOM? How does that work? I understand that the `prepStmtNamespaceAtTxnRewindPos` can keep outstanding references to prepared statements, but it should not *accumulate* them over time - it's merely a snapshot from some point in time.
___
*[pkg/sql/prepared_stmt.go, line 133 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-MiNTQ7ABIbBH8s5UMzc:-MiNTQ7ABIbBH8s5UMzd:b-reytv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/92142a550d4f97509aaad1a6ee0c3f8edef831fe/pkg/sql/prepared_stmt.go#L133)):*
> ```Go
> 	// mutablePreparedPortalState keeps track of the reference counter. This
> 	// needs to be mutable and is, thus, shared among all references.
> 	*mutablePreparedPortalState
> ```

[nit] `refCount *int` is simpler and more clear

**Review by yuzefovich — COMMENTED** [review]
My hypothesis was that `PreparedStatement.decRef` is only called when `PreparedPortal.refCount` reaches 0 which might never happen due to broken refcounting. Initial runs of `schemachange` roachtest were promising, but that was just luck since I still see OOMs.

I refactored the change to remove the refcounting entirely.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-:-MiO39j9AeVmUJg56ljh:b-1w183k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @rafiss)

___
*[pkg/sql/conn_executor.go, line 1469 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-Mi7eU0HsQLYTpdMNDNP-r1-1469:-MiO3X9n71dfQtHTtx_e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6117699bba790bed5b5fed068859b3324b3df341/pkg/sql/conn_executor.go#L1469)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I don't think so, see my explanation in the main thread.
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor.go, line 1476 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-MiNUTu_-5V0z3rNX6qG:-MiO4A2W9Jqs5XT45SN_:b-j2fohf) ([raw file](https://github.com/cockroachdb/cockroach/blob/92142a550d4f97509aaad1a6ee0c3f8edef831fe/pkg/sql/conn_executor.go#L1476)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Is this part what fixes the OOM? How does that work? I understand that the `prepStmtNamespaceAtTxnRewindPos` can keep outstanding references to prepared statements, but it should not *accumulate* them over time - it's merely a snapshot from some point in time.
</blockquote></details>

I think you're right.
___
*[pkg/sql/prepared_stmt.go, line 133 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-MiNTQ7ABIbBH8s5UMzc:-MiO3mjR9vJkIM5YSlfB:b-8gxmw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/92142a550d4f97509aaad1a6ee0c3f8edef831fe/pkg/sql/prepared_stmt.go#L133)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] `refCount *int` is simpler and more clear
</blockquote></details>

I refactored the code to remove the refcounting entirely - I think it no longer makes sense given that we store `PreparedPortal` by value.
___
*[pkg/sql/prepared_stmt.go, line 136 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-Mi7eU0HsQLYTpdMNDNS-r3-136:-MiO3pYo7UWU9Faq0737:b5ol1vw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab453eb532d8186d070a0aed0322a5f98e6d51ff/pkg/sql/prepared_stmt.go#L136)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

thanks for explaining in the PR! that makes me wonder... maybe we could just keep using the `PreparedPortal` by value (and somehow require that it's only used by value). it seems like we only need to care about counting references to `Stmt  *PreparedStatement`. so would things work if we deleted `mutablePreparedPortalState` entirely and made this update instead?

```
-// decRef decrements the number of references to this portal. If the refCount
-// reaches 0, then the memory account is shrunk accordingly.
+// decRef decrements the number of references to the prepared statement used
+// by this portal. The memory account is shrunk accordingly.
 func (p *PreparedPortal) decRef(
     ctx context.Context, prepStmtsNamespaceMemAcc *mon.BoundAccount, portalName string,
 ) {
-    if p.refCount <= 0 {
-        log.Fatal(ctx, "corrupt PreparedPortal refcount")
-    }
-    p.refCount--
-
-    if p.refCount == 0 {
-        prepStmtsNamespaceMemAcc.Shrink(ctx, p.size(portalName))
-        p.Stmt.decRef(ctx)
-    }
+    prepStmtsNamespaceMemAcc.Shrink(ctx, p.size(portalName))
+    p.Stmt.decRef(ctx)
 }
```
</blockquote></details>

I like this idea, done.

**Comment by yuzefovich** [comment]
> i vote that we should still keep this cache disabled, but curious to hear any thoughts on this. i don't feel too strongly, but my reasoning is just that it's not that obvious to a writer of a workload that this statement cache will cause all statements to be prepared. but perhaps that doesn't matter so much.

Removed the second commit as I don't feel strongly either and this PR doesn't seem to fix the OOM.

**Review by RaduBerinde — COMMENTED** [review]
> My hypothesis was that PreparedStatement.decRef is only called when PreparedPortal.refCount reaches 0 which might never happen due to broken refcounting

Right, but the ref count doesn't affect whether the structure gets freed or not - it only affects whether we call `p.memAcc.Close`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69479#-:-MiOIo1O5ECMJNCbOUTe:bq7p3hr)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @rafiss)

**Comment by yuzefovich** [comment]
@RaduBerinde @rafiss do I have your blessing on this change? Thoughts on merging this and backporting it too? I'm kinda on the fence given that it doesn't fix any known problems, but it does fix the memory accounting for the portals - probably leaning towards merging it only after the branch is cut and not backporting.

**Review by rafiss — APPROVED** [review]
i'm +1 on merging this after the branch cut. 

if we see a reason to backport, we can do so at that point

**Comment by RaduBerinde** [comment]
I agree, we should merge on master after the branch cut.

**Comment by RaduBerinde** [comment]
No strong feelings either way for backporting (other than waiting for a bake-in period on master).

**Comment by yuzefovich** [comment]
TFTRs!

Will merge after the branch is cut and will not backport.

**Comment by yuzefovich** [comment]
The branch has been cut, and I rebased on top of latest master just in case.

bors r+

**Pull request merged**
